articles/best-practices/rules.md

```diff
@@ -1,247 +1,565 @@
+:::
+
+The first two objects passed to the rule during testing represent the [`user`](#user-object) and [`context`](#context-object), and these can be mocked in a fashion similar to that employed in the Auth0 Dashboard `TRY` functionality. The [`callback`](#callback-function) function, supplied as the third parameter, can be implemented to simulate pipeline continuation, subsequently performing execution of the next rule in [order](#order). 
+
+::: panel Best Practice
+The `callback` function supplied can be used to ensure execution of callback is performed *at least* once by it (the function) executing test of the next rule in the chain. Implementing test(s) in the supplied function to also ensure that multiple execution of the callback is not performed by a rule is also a recommended best practice.
+:::
+
+In addition, the (Node.js) `global` object can be used to provide both the [configuration](#environment) object, and also an instance of the [`auth0`](#auth0-object) object if required. In the sample above, a global `configuration` object has been defined that aligns with recommended practices to assist with [debugging](#debugging) (as described in the section above) .
+
+The sample above also makes use of the file system directory structure provided by Auth0 [Deploy CLI](/extensions/deploy-cli) - the tooling which can assist with rule [deployment](#deployment), as described in the section below.  
+
+## Deployment
+Coding a rule within the Auth0 Dashboard rule editor is a great way to implement and test whilst still in the development stage. However, when it comes time to deploy into automated test and/or production environments, a more automated mechanism is required; copy and pasting code between Auth0 tenants is not a satisfactory method to employ.
+
+Out of the box, Auth0 provides a number of facilities for automated deployment of rule extensibility assets between Auth0 tenant environments. The Auth0 [GitHub](/extensions/github-deploy), [GitLab](/extensions/gitlab-deploy) and [Bitbucket](/extensions/bitbucket-deploy) extensions provide the ability to update rule assets from the respective 3rd party version control system - both manually, and in many instances automatically too (i.e. when a change in the version control system is detected).
+
+In addition, the Auth0 [Deploy CLI](/extensions/deploy-cli) tool can be used to automate deployment between Auth0 tenants. Deploy CLI works with files stored in the file system together with the Auth0 Management API, and provides capability to allow the export of rule assets from an Auth0 tenant, as well as import of them into an Auth0 tenant. Further the tool provides for programmatic control over rule ordering and rule environment [configuration](#environment-variables), as part of deployment automation. In many ways, the Deploy CLI is like a Swiss Army Knife when it comes to rule extensibility deployment in Auth0.  
```

Comment:
Your re-factor seems to suggest that using the Auth0 Extensions as part of a CI/CD pipeline is acceptable practice. Whilst there's nothing to prevent a customer from doing so, I would argue that it's going to make life a whole less flexible - and potentially harder - if you **don't** use the CLI tooling instead (i.e not just as a good alternative). 

For example, customers may want to add static analysis checking of their rules as part of CI/CD pipeline workflow, and possible gate deployment to an Auth0 tenant (e.g. via something like Gerrit). And that's going to be tricky to do if you use the Extension, as push deployment via extension will trigger *before* analysis tooling is run in the pipeline.

Reply:
For now I'm going to leave the statement as it stands